# <br><br><span style="color:rebeccapurple">Python Fundamentals Bootcamp - Day 1</span>

### <br><br>This is a Jupyter Notebook. 
To run a gray code cell, click in the cell and either click on the "play" arrow, or type shift+enter (or shift+return on a Mac).

In [1]:
print("Hello World!")

Hello World!


<br>When you type in the gray code cell and hit shift+enter, you **run** the code.
<br>If something gets printed out to the screen after you run the code, we say it is **returned**.

Text cells like this one don't contain code. They are written in a language called Markdown. If you accidentally double click in a markdown cell, it won't cause any problems. Go ahead and double click on this text. Then you can run the cell, just like you would with a code cell by hitting shift+enter, to return it to looking like nice formatted text.

<br>Everyone's notebook might look a little different, depending on if you running it locally through Jupyter Lab, on Google Colab, on a Mac or a PC, or in different browsers.

<br>As we work, you might feel like you need to reset your entire notebook. I don't expect you to remember this, but I'm putting this info here in case you need it later:
- In Jupyter Lab: Click on Kernel in the menu to view options
- In Google Colab: Click on Runtime in the menu to view options

<br>*If you are working on Google Colab:* Some of the groups of cells will automatically be **collapsed**, requiring you to **expand** them as you go. If you want, you can click on the View menu and choose Expand sections to expand the entire notebook.

#### <br>Before we start coding, here are some things you should know that will make Python easier to learn...

## <br><br><span style="color:teal">Python is a language.

(Not a software program.)

#### How do we talk to our computer in Python?

- **Interactive programming**: through a **shell** (interpreter, console) one line at a time
- **Batch programming**: running a whole **script** (a plain text file containing one to many lines of code)
- With the help of a GUI (graphical user interface). GUIs for coding are called **IDEs (Integrated Development Environments)**.
    - Some Python IDES: Spyder, PyCharm, Jupyter Lab, Google Colab

*The Jupyter Notebook we're working in now isn't exactly an IDE - it can be used for both interactive (one line at a time) and batch programming (many lines at a time).*

## <br><br><span style="color:teal">Python is:

**50% Syntax**
- Which words to use
- Punctuation
- Order
- Indentation
- Shortcuts

**50% Logic**
- Which tools to use in which order
- Creative
- Specific to one problem

<br>Try not to worry if you don't understand everything today. You'll get more practice and time to think (and overthink!) on the quiz. I will label some concepts we cover as **<span style="color:crimson">LOGIC**. LOGIC skills are very helpful for piecing together your own code, but if you are struggling on a LOGIC section, don't worry too much - you can focus on learning the syntax and come back to thinking about logic when you are writing your own code.

#### <br>Python is a very different language than R.
If you are an R coder, Python works very differently. R was originally written for statistical analysis and is primary a *functional* language. This means that you are often thinking "How do I do what I want to do?" and then look for the function or series of functions you will apply to your data. Python was originally written for software development and is primarily an *object-oriented* language. This means that you should think "What exactly do I want to do?" and then write code to do exactly what you want instead of looking for predefined functions.
<br><br>The power of Python comes from lists, dictionaries, loops, and conditionals. Python is best for anything that comes close to software development, including automating tasks, complicated logic problems, coding new computational methods, and creating completely custom solutions. 

#### <br>Python is a language, but you'll never have to speak it in front of anyone!
You do not have to memorize everything you learn - Coders of all levels rely on searching the internet to help them recall the syntax for a particular task. Your search engine will point you to the official Python documentation or forums like Stack Overflow or blog tutorials.

## <br><br><span style="color:teal">Objects and Functions

#### The two main concepts of Python.
- **Object**: a particular piece of data (like a noun)
- **Function**: something you can do to/with an object (like a verb)

<br>You can use certain functions with certain classes of objects.
<br>
<br>Some functions are shared between classes of objects, and some are not.

<br><br><br>Let's consider this example. I have a dog named Molly, and I have a bicycle. There are some things I can do to both Molly and my bicycle, and some things I can't do to both:

## 🐕 Molly

- Feed Molly
- Pet Molly
- Draw Molly

## 🚲 My bicycle
- Pedal my bicycle
- Pet my bicycle (limited use)
- Draw my bicycle

<br><br><br>The things I can do to Molly are actually things I can do to any dog, and the things I can do to my bicycle are things I can do to any bicycle:

## 🐕 Molly (class dog)

- Feed a dog
- Pet a dog
- Draw a dog

## 🚲 My bicycle (class bicycle)
- Pedal a bicycle
- Pet a bicycle (limited use)
- Draw a bicycle

#### <br><br>*Note: Object Classes are the exact same as Data Types in Python.*

<br><br>Python example: We can do different things to the number three, depending on which type of object it is.

## 🔢 &nbsp; 3 (class integer)

- Divide (3 / 3 = 1)
- Convert 3 to 3.0
- Add (3 + 3 = 6)

## 📝 &nbsp; "3" (class string)
- Replace ("3" becomes "1")
- Convert "3" to 3.0
- Add ("3" + "3" = "33")

### <br><br><br>Today's Objects
- Integer (whole number) 3
- Float (number with digits following a decimal point) 3.0
- String (text or characters) “3” or “three”
- Boolean (True or False)
- *Assigning objects to variables*
- *Indexing to get only part of an object*

### Today's Functions
- Basic operators: + - * /
- Functions to convert between data types
- String functions

## <br><br><br><span style="color:magenta">**LET'S CODE!**

## <br>Arithmetic operators

In [2]:
4 + 2

6

In [3]:
6 - 1

5

In [4]:
100 * 35

3500

In [5]:
40 / 6

6.666666666666667

To calculate an exponential:

In [6]:
4 ** 2

16

<br><br>You might notice that I've included spaces around the operators: 
<br>`4 + 2`
<br>instead of
<br>`4+2`
<br><br>This is best practice in Python because it is more readable. The math would work without spaces, but it's a good idea to get into the habit of writing "Pythonic" code now!

#### <br><br>More division operators:

In [7]:
40 / 5

8.0

We gave it **integers** (whole number objects) and it returned a **float**. The `/` division operator always returns a float.

<br>Get just the integer, without the remainder:

In [8]:
40 // 6

6

Get just the remainder:

In [9]:
40 % 6 

4

<br><span style="color:crimson">Like many things you'll learn this week, I'm not expecting you to memorize the operator for returning a remainder. If this is something you use a lot already in your research, you might memorize it now. But if not, I just want you to hear about it so that if you ever need it months from now, you might remember that there is a way to do it so that you can google it and find the right code.</span>

<br>More practice:

In [10]:
40 / 5

8.0

In [11]:
40 // 5

8

In [12]:
40 % 5

0

#### <br><br>Python follows the PEMDAS order of operations: (https://thehelloworldprogram.com/python/python-operators-order-precedence/)

In [13]:
40 / 5 + 2

10.0

In [14]:
40 + 2 * 6 - 1

51

In [15]:
(40 + 2) * (6 - 1)

210

Why did some lines of code return integers (51, 210) and one returned a float (10.0)? <br><br><br><br>The first line of code included division `/`, which you may remember always returns a float.

### <br><br><span style="color:red">Exercise: Arithmetic Operators

Try out some math of your own:

If you ever want to add another code cell to you notebook, you can. Click in the cell *above* where you want to add a cell, and then click the **+** sign in the menu at the top of the page. (On Google Colab it will say **+ Code**). Let's try it now. Click in this cell and then click on the **+** button above (or **+ Code**).

## <br><br>Comparison operators

These return a **boolean** object: True or False.

In [16]:
3 > 2

True

In [17]:
3 < 2

False

In [18]:
3 >= 2

True

In [19]:
3 <= 2

False

To check if two things are **equal**, Python uses a double equals sign:

In [20]:
3 == 3

True

To check if two things are **not equal**, Python uses `!=`:

In [21]:
3 != 3

False

In [22]:
3 != 2

True

### <br><br><span style="color:red">Exercise: Comparison Operators

Write code to find out if 1.00486 is greater than 1.000918.

In [23]:
1.00486 > 1.000918

True

Write code to find out if 6 is equal to 6.0.

In [24]:
6 == 6.0

True

## <br><br>Logical operators

These also return a **boolean**.

`and` means both sides must be True to return True:

In [25]:
3 < 4 and 4 == 3

False

`or` means only one side must be True to return True:

In [26]:
3 < 4 or 4 == 3

True

`not` is another logical operator. It reverses the boolean that follows it:

In [27]:
not 2 == 2

False

### <br><br><span style="color:red">Exercise: Logical Operators

Before running these cells, try to guess if they will return True or False:

In [28]:
not 2 > 3 and not 2 == 2

False

In [29]:
not 2 > 3 or not 2 == 2

True

In [30]:
3 == 3 and not 3 != 3

True

## <br><br>String objects

A **string** is an object kind of like text. 
<br><br>It is contained in double `"` or single `'` quotes (you can choose which you want to use, but I use double quotes, so that is what we will use for this bootcamp).

In [31]:
"Hello World"

'Hello World'

<br><br>Let's use our first **function**! The `print()` function can be used on almost all classes of objects:

In [32]:
print("Hello World")

Hello World


<br>Strings can contain digits:

In [33]:
print("Hello World 2")

Hello World 2


#### <br><br>Special Characters

Strings can contain **special characters**. These start with a backslash (on my keyboard, the backslash key is found above my return key). The backslash can also be called the **escape character**. <br><br>The most commonly used special characters are new lines `\n`:

In [34]:
print("Hello World\nHow are you?")

Hello World
How are you?


<br>...and tabs `\t`:

In [35]:
print("Hello World\tHow are you?")

Hello World	How are you?


<br><br><br>If you simply **run** the string, it will return the string how it looks to the computer:

In [36]:
"Hello World\nHow are you?"

'Hello World\nHow are you?'

<br>...but if you use the `print()` function, it **interprets** the special characters and returns the string like this:

In [37]:
print("Hello World\nHow are you?")

Hello World
How are you?


<br><br><br>You can include single and double quotation marks in your strings:

In [38]:
print("Here's an example.")

Here's an example.


In [39]:
print('Hello "World"')

Hello "World"


In the first example, we used a single quotation mark (apostrophe) inside double quotation marks. In the second example, we used double quotation marks inside single quotation marks.

<br>What if you want to include both single and double quotation marks in a string? You can use the escape character (the backslash) to let the computer know to interpret the following character as part of the string:

In [40]:
print("Here's the \"example\".")

Here's the "example".


<br>If you ever need to include an actual backslash in your string and you don't want it to be interpreted as a special character, you can use a double backslash:

In [41]:
print("always\never")

always
ever


In [42]:
print("always\\never")

always\never


### <br><br><span style="color:red">Exercise: Strings and Special Characters

Write code to print out the following message (it should look exactly as it is shown here):
<br><br>This is line 1.
<br>This is line 2.

In [43]:
print("This is line 1.\nThis is line 2.")

This is line 1.
This is line 2.


#### <br><br><br>Python errors

What happens if you forget the quotes around a string?

In [44]:
print(Hello World)

SyntaxError: invalid syntax (4293340409.py, line 1)

<br><br>Our first error!! This is a common error - the `SyntaxError`. It often means that we forgot to type something.

<br><br>What happens if you spell a function wrong?

In [45]:
pint("Hello World!")

NameError: name 'pint' is not defined

<br><br>A `NameError` is another common error - it often means we've mispelled a function or variable.

<br><br>What error do you think we will get if we run this line of code?

In [46]:
print("Hello World)

SyntaxError: EOL while scanning string literal (2771583741.py, line 1)

#### <br><br>Combining strings

Some arithmetic operators work with strings. To combine two strings:

In [47]:
print("Hello" + "World")

HelloWorld


In [48]:
print("Hello " + "World")

Hello World


## <br><br>Converting between data types

<br>The `type()` function tells us the data type of an object (in Python, a data type is exactly the same thing as an object class):

In [49]:
type("Hello World")

str

In [50]:
type(9)

int

In [51]:
type(9.9)

float

In [52]:
type("9.9")

str

<br>We can convert objects from one type to another, using the appropriate function:

In [53]:
float(9)

9.0

In [54]:
float("5")

5.0

In [55]:
int("5")

5

In [56]:
int(5.7)

5

*Notice it doesn't round when it converts a float to an integer, it just drops off everything after the decimal. You can use the `round()` function to round, but we'll cover that a bit later.*

In [57]:
str(5.005)

'5.005'

In [58]:
float("Hello World")

ValueError: could not convert string to float: 'Hello World'

<br>Our third type of error! This error usually means you passed an argument to a function that it doesn't recognize. But don't spend any brainpower trying to memorize this error. Luckily the error message gives us information about what we did wrong.

## <br><br>Data types interacting

Before running each of these cells, try to guess what they will return:

In [59]:
4 * 4.0

16.0

In [60]:
int(4 * 4.0)

16

In [61]:
"Hello World" * 4

'Hello WorldHello WorldHello WorldHello World'

In [62]:
5 + True

6

In [63]:
5 + False

5

*Note: True evaluates to 1, False evaluates to 0*

### <br><br><span style="color:red">Exercise: Data Types Interacting

Run the next line of code:

In [64]:
"Hello World" + 4

TypeError: can only concatenate str (not "int") to str

<br>A TypeError means you tried to do something that doesn't work with the type of object you gave it. Sometimes ValueErrors and TypeErrors get thrown for similar errors.

<br>Use a function to modify the code below to create a string that says "Hello World4".

In [65]:
print("Hello World" + str(4))

Hello World4


In [66]:
print("Hello World" + "4")

Hello World4


## <br><br><br>Variables

Variables allow you to store an object to use it later.

In [67]:
x = 3

Nothing is returned when you assign a variable. You can see the value of a variable by running the variable name or using the `print()` function

In [68]:
x

3

In [69]:
print(x)

3


<br>The single `=` is called an **assignment operator**.
<br>Everything on the right side of the assignment operator is calculated first, and then that value is stored as the variable on the left of the assignment operator.

In [70]:
x = 3 - 1

In [71]:
x

2

In [72]:
y = x

In [73]:
y

2

In [74]:
x = x**y

In [75]:
x

4

In [76]:
y

2

In [77]:
rabbit = 50

Variables can be used just like the objects they represent:

In [78]:
rabbit - 20

30

In [79]:
rabbit

50

In [80]:
greeting = "Hello you!"

In [81]:
greeting

'Hello you!'

In [82]:
print(greeting)

Hello you!


### <br><br><span style="color:red">Exercise: Creating Variables

Create a variable called `favorite_color` and assign it to a string that contains your favorite color. Then write code to print the string you saved as `favorite_color`.

In [84]:
favorite_color = "navy blue"

In [85]:
print(favorite_color)

navy blue


In [86]:
str(blue)

NameError: name 'blue' is not defined

## <br><br>Variable names

#### Python Rules
- No spaces
- Case matters: x is not X
- Must start with a letter
- Should be meaningful (don't save your number as `rabbit` unless it represents the number of rabbits)
- For longer variable names: separate_with_underscores

## <br><br><span style="color:teal">Functions

We've been learning a lot about objects. Let's learn more about functions!

<br>**Parentheses immediately follow a function name (no space).**

In [87]:
print("Hello World")

Hello World


The information inside the parentheses is called an **argument**.
<br>We say we **passed** the argument "Hello World" to the print function.

### <br><br>Two types of functions in Python

1. A true **function**. 
- Stands on its own. 
- Takes an object as an **argument**. 
- These types of functions usually do something **with** the object.

In [88]:
my_name = "Colby"

In [89]:
print(my_name)

Colby


In [90]:
len(my_name)

5

<br><br>2. **Method**
- Always follows an object. 
- Some take arguments and some don't. 
- These types of functions usually do something **to** the object. 

In [91]:
my_name = "Colby"

In [92]:
my_name.upper()

'COLBY'

In [93]:
my_name.replace("C", "c")

'colby'

<br><br>When a function takes multiple arguments, they are separated by a comma:

In [94]:
my_name.replace("C", "c")

'colby'

In [96]:
round(4.5874, 1)

4.6

### <br><br><span style="color:red">Exercise: String variables

Create a new variable called `full_name`. It should store the string of your own full name.

In [97]:
full_name = "Colby Witherup Wood"

In [98]:
print(full_name)

Colby Witherup Wood


Write code to find out the length of your full name.

In [99]:
len(full_name)

19

Did the length function count spaces when it calculated the length of your name?

## <br><br>String functions

In [100]:
greeting = "Hello you!"

In [101]:
print(greeting)

Hello you!


In [102]:
len(greeting)

10

In [103]:
greeting.replace("!", ".")

'Hello you.'

In [104]:
greeting.replace(" ", "")

'Helloyou!'

`""` is referred to as an **empty string**.

In [105]:
print(greeting)

Hello you!


Even though we ran code to change our greeting, the greeting variable still **points** to the original greeting. `greeting` did not change when we used the method functions even though a changed version of `greeting` was **returned**.

<br>To change the greeting, we would have to **reassign** the variable name:

In [106]:
greeting = greeting.replace("!", ".")

In [107]:
print(greeting)

Hello you.


In [108]:
big_greeting = greeting.upper()

In [109]:
print(big_greeting)

HELLO YOU.


In [110]:
small_greeting = greeting.lower()

In [111]:
print(small_greeting)

hello you.


In [112]:
my_name = "Colby"

In [113]:
greetMe = greeting.replace("you", my_name)
print(greetMe)

Hello Colby.


*Note: You might have noticed that I included two lines of code in that code cell. Jupyter Noteboooks can run multiple lines of code in a single code cell.*

<br>You will learn more string functions and get practice with string functions in the afternoon session.

## <br><br><span style="color:teal">Indexing strings

We can **index** strings to return a **substring**. A substring can be one or more characters long.

#### <br><span style="color:magenta">**In Python indexing, we start counting with 0.**

#### H &nbsp; &nbsp; E &nbsp; &nbsp; L &nbsp; &nbsp; L &nbsp; &nbsp; O &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; W &nbsp; &nbsp; O &nbsp; &nbsp; R &nbsp; &nbsp; L &nbsp; &nbsp; D &nbsp; &nbsp; !
#### 0 &nbsp; &nbsp; 1 &nbsp; &nbsp; 2 &nbsp; &nbsp; 3 &nbsp; &nbsp; 4 &nbsp; &nbsp; 5 &nbsp; &nbsp; 6 &nbsp; &nbsp; 7 &nbsp; &nbsp; 8 &nbsp; &nbsp; 9 &nbsp; &nbsp; 10 &nbsp; &nbsp; 11

<br>To return a single character, we type our string followed immediately by square brackets (no space). Inside the brackets, we pass the index position for the character.

In [114]:
"Hello World!"[0]

'H'

In [115]:
"Hello World!"[4]

'o'

#### <br><br>H &nbsp; &nbsp; E &nbsp; &nbsp; L &nbsp; &nbsp; L &nbsp; &nbsp; O &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; W &nbsp; &nbsp; O &nbsp; &nbsp; R &nbsp; &nbsp; L &nbsp; &nbsp; D &nbsp; &nbsp; !
#### 0 &nbsp; &nbsp; 1 &nbsp; &nbsp; 2 &nbsp; &nbsp; 3 &nbsp; &nbsp; 4 &nbsp; &nbsp; 5 &nbsp; &nbsp; 6 &nbsp; &nbsp; 7 &nbsp; &nbsp; 8 &nbsp; &nbsp; 9 &nbsp; &nbsp; 10 &nbsp; &nbsp; 11

To take a longer **substring**, you start with the position of the first character you want included, then a colon, and end with the position *1 past* the last character you want.

To index "Hello":

In [116]:
"Hello World!"[0:5]

'Hello'

<br><br>If this seems strange, it might help you to know what is happening behind the scenes. Python is actually labeling the invisible spot *between* each character in our string:

#### &nbsp; &nbsp; H &nbsp; &nbsp; E &nbsp; &nbsp; L &nbsp; &nbsp; L &nbsp; &nbsp; O &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; W &nbsp; &nbsp; O &nbsp; &nbsp; R &nbsp; &nbsp; L &nbsp; &nbsp; D &nbsp; &nbsp; !
#### 0 &nbsp; &nbsp; 1 &nbsp; &nbsp; 2 &nbsp; &nbsp; 3 &nbsp; &nbsp; 4 &nbsp; &nbsp; 5 &nbsp; &nbsp; 6 &nbsp; &nbsp; 7 &nbsp; &nbsp; 8 &nbsp; &nbsp; 9 &nbsp; &nbsp; 10 &nbsp; &nbsp; 11 &nbsp; 12

So now it's easier to see why "Hello" would be indexed with [0:5].<br><br><br><br>

In [117]:
greeting = "Hello you!"

In [118]:
greeting[0:5]

'Hello'

In [119]:
hello = greeting[0:5]
print(hello)

Hello


In [120]:
hello[2]

'l'

*Remember, in Python we start indexing at 0, so the third letter in our string is indexed as 2.*

<br><br>We can also start indexing from the right end of the string. The character in the farthest right position can be indexed as -1. We call this **negative indexing**.

In [121]:
hello[-1]

'o'

In [122]:
hello[-2]

'l'

In [123]:
hello[0:2]

'He'

*Remember, our variable is `hello`, but the string it has stored is "Hello" with a capital "H".*

<br><br>If we don't give any number after the `:`, Python will continue to the end of the string:

In [124]:
hello[3:]

'lo'

If we don't give any number before the `:`, Python will start at the beginning of the string:

In [125]:
hello[:2]

'He'

<br>We can also use negative numbers to index substrings:

In [126]:
hello[2:-1]

'll'

In [127]:
hello[-3:-1]

'll'

In [128]:
"Hello"[-3:]

'llo'

### <br><br><span style="color:red">Exercise: Index a string

Run the line of code below to store the variable `movie`.

In [130]:
movie = "TENET"

Write code to index the first character in the variable `movie`.

In [131]:
movie[0]

'T'

Use negative indexing to return the last character in the variable `movie`.

In [132]:
movie[-1]

'T'

**Bonus exercise**
<br>Is the first letter of `movie` the same as the last letter of `movie`? Combine the indexing statements you just wrote with a comparison operator to see if the two letters are equal.

In [133]:
movie[0] == movie[-1]

True

*Your code should return a boolean: True or False.*

#### <br><br>One final concept to learn this morning...

## <br><span style="color:teal">Modules (aka packages or libraries)

By default, only a handful of the most commonly used functions and object classes are loaded when you start Python.
<br><br>This saves memory so that your Python scripts can run as quickly as possible.
<br><br>The object classes and functions that are loaded automatically with Python are called **built-in**.

#### <br>Modules
- Modules are groups of related functions and object classes
- They are not automatically loaded when you start Python
- You must **import** the module every time you start Python (or in every Jupyter Notebook)
- If you do not have the module on your computer, you must first **install** it (only once)

#### <br>Where to find modules
- Anaconda comes with many of the most common modules pre-installed (we'll talk more about Anaconda on Day 3)
- Many other modules can be installed from the command line
- Some must be downloaded from the creator
- There are modules for every field and use

<br>Today we'll learn how to **import** packages and work with a couple functions from a package that is included in every installation of Python but isn't automatically loaded when you start Python.
<br><br>Tomorrow we'll talk briefly about **installing** packages.

We are going to use some functions from the module `math`.

You should usually import any modules at the top of your Jupyter notebook, even if you don't use them until much later in the notebook. This is so that people can see which modules they need as soon as they open the notebook, in case they need to **install** the module on their computer. *We're breaking this rule today and loading the module here.*

Run this cell to import the `math` module:

In [134]:
import math

<br><br>To use a function from an imported module, you type the name of the module followed by `.` followed by the name of the function. Let's use `math.sqrt()` to find the square root of a number:

In [135]:
math.sqrt(16)

4.0

In [136]:
math.sqrt(15.5)

3.9370039370059056

<br>The `math` module contains many mathematical functions. It also contains a few objects, like `math.pi`, which is the value of pi, so that you don't have to memorize it. You know it's an object and not a function because it doesn't have parentheses at the end.

In [137]:
radius = 6
area_of_a_circle = math.pi * radius**2
print(area_of_a_circle)

113.09733552923255


*Remember that `**` is used to find an exponent like a square.*

### <br><br><span style="color:red">Exercise: Use an object from a module

The equation for calculating the circumference of a circle is 2 x pi x radius. Write code to calculate the circumference of a circle with a radius of 10.

In [138]:
radius = 10
circumference = 2 * math.pi * radius
print(circumference)

62.83185307179586


# <br><br><span style="color:rebeccapurple">LUNCH BREAK - meet back at 1 pm Central

**If you're working in Colab**, you might need to refresh your page. You'll lose the work you did above - but if you'd like to save it you can either: <br>1. Click the button at the top Copy to Drive and save it in your personal Google Drive, or
<br>2. Go to File>Download>Download .ipynb

## We'll get started at 1:02 pm Central.

## <br><br><span style="color:teal">More fun with integers and floats
### <br><br>Let's turn some story problems into Python code.

*When running this code, don't forget to run every code cell. If you get a `NameError` it might be because you didn't run a cell where a variable was defined.*

<br>I told the class I ate 12 bananas for lunch:

In [139]:
bananas = 12

You saw me secretly eat 3 more bananas:

In [140]:
bananas = bananas + 3

How many bananas did I eat?

In [141]:
print(bananas)

15


Let's add some strings together to print the answer out as a complete sentence:

In [142]:
answer = "Colby ate " + bananas + " bananas."

TypeError: can only concatenate str (not "int") to str

<br>Error! Let's check the type of the variable `bananas`.

In [143]:
type(bananas)

int

We can change `bananas` to a string and save it as a new variable so that we can include it in our answer...

In [144]:
bananas_str = str(bananas)
answer = "Colby ate " + bananas_str + " bananas."
print(answer)

Colby ate 15 bananas.


<br>...or we can change it to a string *inside* the code to create our answer:

In [145]:
answer = "Colby ate " + str(bananas) + " bananas."
print(answer)

Colby ate 15 bananas.


<br><br>Our final code for this story problem looked like this:

In [146]:
bananas = 12
bananas = bananas + 3
answer = "Colby ate " + str(bananas) + " bananas."
print(answer)

Colby ate 15 bananas.


<br>I also could have written the code like this, creating a new variable in the second line instead of reusing the variable name `bananas`:

In [147]:
bananas = 12
real_bananas = bananas + 3
answer = "Colby ate " + str(real_bananas) + " bananas."
print(answer)

Colby ate 15 bananas.


<br>Or I could have written the code like this:

In [148]:
bananas = 12
secret_bananas = 3
real_bananas = bananas + secret_bananas
answer = "Colby ate " + str(real_bananas) + " bananas."
print(answer)

Colby ate 15 bananas.


<br>Or I could have written the code like this:

In [149]:
bananas = 12
print("Colby ate " + str(bananas + 3) + " bananas.")

Colby ate 15 bananas.


<br>These alternative ways of getting to the same answer are all acceptable. The difference between them involves when I decided to assign values to variables. There isn't always a right or wrong answer, but there are some guidelines to make your code more "Pythonic".

<br>In general, you want your code to be succinct but readable. You don't want to store too many variables in your computer's memory at one time, especially if they contain large objects like whole files (we'll talk about files and memory on Day 3). If you will never use the object again in the code, you might not want to bother assigning that object to a variable. Then again, if it makes it more readable to assign it to a variable, you should.

<br>For now, just focus on writing code that works! All of the options above would be acceptable.

### <br><br><span style="color:red">Integers and Floats Exercise 1

My kitchen is 10 feet long by 10 feet wide. How many square feet is my kitchen? (The equation for area is length multiplied by width.) Finish the code below by assigning the area equation to the variable `area`:

In [150]:
length = 10
width = 10
area = length * width
print("The kitchen is " + str(area) + " square feet.")

The kitchen is 100 square feet.


### <br><br><span style="color:red">Integers and Floats Exercise 2

I scored 80 points on my first test and 88.5 points on my second test. What is my average test grade? (The equation for average test grade is the sum of all test scores divided by the number of tests.) Fill in the appropriate code below:

In [151]:
test1 = 80
test2 = 88.5
average = (test1 + test2) / 2
print("Colby's average test grade is " + str(average) + ".")

Colby's average test grade is 84.25.


### <br><br><span style="color:red">Integers and Floats Exercise 3

Last week I ran 10 miles. This week I ran the same as last week, plus 6 extra miles. How many miles did I run in total over both weeks? Write the code below to calculate the total number of miles. As a bonus, write the code to print out a complete sentence that says how many miles I ran. 

In [152]:
week1 = 10
week2 = week1 + 6
total = week1 + week2
print("In two weeks, I ran " + str(total) + " miles.")

In two weeks, I ran 26 miles.


In [153]:
week1 = 10
week2 = week1 + 6
print("In two weeks, I ran " + str(week1 + week2) + " miles.")

In two weeks, I ran 26 miles.


### <br><br><br>Let's learn some new functions that work with integers and floats.

<br>The `abs()` function returns the absolute value of a number:

In [154]:
abs(-19.6)

19.6

<br>The `round()` function will round a float to a specified number of decimal places. It takes two arguments: the number to round and the number of decimal places.

In [155]:
my_number = 631.890382720

In [156]:
round(my_number, 4)

631.8904

In [157]:
round(my_number, 2)

631.89

Some function arguments have **default values**, so if you don't include the argument, it will use the default value. Let's try `round()` without specifying a value for the second argument:

In [158]:
round(my_number)

632

So the default value is to round to the closest integer.

`round()` can also round to places to the left of the decimal point:

In [159]:
round(18192948, -3)

18193000

<br>You can also put functions inside other functions:

In [160]:
round(abs(-12.908))

13

In [161]:
abs(round(-12.908))

13

In [162]:
round(30 + 698, -1)

730

### <br><br><span style="color:red">Integers and Floats Exercise 4

I am ordering 3 tacos for lunch. Each taco is $2.55. I would like to round my total up to the nearest dollar to leave a small tip. How much should I pay? Write the code below:

In [163]:
taco = 2.55
round(taco * 3)

8

<br><br><br>Some functions have more than one argument that has a default value. To change the default value, you have to pass the function a **keyword argument**. Let's see how this works.

The function `isclose()` is in the `math` module. Let's import the math module again, in case you closed and reopened this notebook during lunch.

In [164]:
import math

<br><br>`math.isclose()` will tell you if two numbers are close. It returns a boolean.

In [165]:
math.isclose(4, 5)

False

4 and 5 are not considered close.

In [166]:
math.isclose(0.4, 0.5)

False

Still not close.

In [167]:
math.isclose(0.0004, 0.0005)

False

Still not close.

In [168]:
math.isclose(.4999999999, .5)

True

*Relative tolerance* is the maximum allowed difference between a and b. The default relative tolerance is 1e-09.
<br><br>We can change the relative tolerance by passing the keyword argument `rel_tol`.

In [169]:
math.isclose(4.03, 4.02, rel_tol=.01)

True

<br><br>We can get information about a function by looking it up in the documentation <https://docs.python.org/3/library/index.html> or by coding the function name without the `()` followed by `?`:

In [170]:
math.isclose?

Signature: math.isclose(a, b, *, rel_tol=1e-09, abs_tol=0.0)
Docstring:
Determine whether two floating point numbers are close in value.

  rel_tol
    maximum difference for being considered "close", relative to the
    magnitude of the input values
  abs_tol
    maximum difference for being considered "close", regardless of the
    magnitude of the input values

Return True if a is close in value to b, and False otherwise.

For the values to be considered close, the difference between them
must be smaller than at least one of the tolerances.

-inf, inf and NaN behave similarly to the IEEE 754 Standard.  That
is, NaN is not close to anything, even itself.  inf and -inf are
only close to themselves.
Type:      builtin_function_or_method


## <br><br><span style="color:teal">More fun with strings

<br>We already learned these string functions:

In [171]:
song = "We are the world"

In [172]:
song.lower()

'we are the world'

In [173]:
song.upper()

'WE ARE THE WORLD'

In [174]:
song.replace("world", "weird")

'We are the weird'

In [175]:
new_word = "wind"
song.replace("world", new_word)

'We are the wind'

<br>We also learned that these functions don't change the song:

In [176]:
print(song)

We are the world


...and that we can save a changed version of the object by assigning a new variable, or reassigning the old variable:

In [177]:
song = song.upper()
print(song)

WE ARE THE WORLD


### <br><br>More capitalization functions

Run the code below to see what the `title()` and `capitalize()` methods do.

In [178]:
song = "All along the watchtower"

In [179]:
song.title()

'All Along The Watchtower'

In [180]:
song = song.lower()
print(song)

all along the watchtower


In [181]:
song.capitalize()

'All along the watchtower'

### <br><br>Removing characters from a string

In [182]:
print(song)

all along the watchtower


<br>The `lstrip()` method will remove a character or set of characters if it is the first character on the left side of the string:

In [183]:
song.lstrip("a")

'll along the watchtower'

In [184]:
song.lstrip("all")

' along the watchtower'

In [185]:
song.lstrip("w")

'all along the watchtower'

"w" is not the first character on the left side of the string, so nothing was removed.

<br><br>The `rstrip()` method will remove a character or set of characters if it is the first character on the right side of the string:

In [186]:
song.rstrip("er")

'all along the watchtow'

<br>I use `rstrip()` all the time to remove new line characters from the end of lines in a file, or to remove trailing whitespace:

In [187]:
"line in a file\n\n".rstrip("\n")

'line in a file'

In [188]:
"line in a file      ".rstrip(" ")

'line in a file'

*Notice that it removed all the spaces on the right end of the string, even though we only passed it one space as the argument.*

<br><br>**If you want to remove a character everywhere it appears in a string, you should use the `replace()` method we already learned and replace the character with an empty string**:

In [189]:
song.replace(" ", "")

'allalongthewatchtower'

### <br><br><span style="color:red">Strings Exercise 1

Run the next line of code to save the variable `book`.

In [191]:
book = "The Three Musk-eteers"

Write code to remove the "-" in the middle of `book`.

In [192]:
book.replace("-", "")

'The Three Musketeers'

**Bonus exercise:** Write code to remove the "-" in the middle of the word "Musketeers" in `book2`, but not the "-" that divides the title from the subtitle.

In [193]:
book2 = "The Three Musk-eteers - Part Two"

In [195]:
book2.replace("Musk-", "Musk")

'The Three Musketeers - Part Two'

In [196]:
book2.replace("-", "", 1)

'The Three Musketeers - Part Two'

### <br><br>Stringing together method functions

You can string multiple method functions on the end of an object. The functions will be executed from left to right.

In [197]:
song = "Let It Go"

In [198]:
different_song = song.replace("Go", "Be").upper()
print(different_song)

LET IT BE


<br>The order of functions can change what is returned. Why do you think reversing the order of functions turns out the way it does?

In [199]:
different_song = song.upper().replace("Go", "Be")
print(different_song)

LET IT GO


<br><br><br>Once you've made the song uppercase, "Go" does not exist in the string, so it won't be replaced. "Go" is already "GO".

<br>*Side note, in case this ever happens to you, if you forget to put the parantheses on the end of your method function, you might not get an error, but you will see something like this:*

In [200]:
song.lower

<function str.lower()>

### <br><br>Indexing and function practice

In this morning's lecture you learned how to index a string to return a substring. 

In [201]:
song_line = "Row, row, row your boat"
what_to_row = song_line[-9:]
print(what_to_row)

your boat


In [202]:
what_to_do_with_your_boat = song_line[:3]
print(what_to_do_with_your_boat)

Row


In [203]:
whose_boat_it_is = song_line[14:18]
print(whose_boat_it_is)

your


### <br><br><span style="color:red">Strings Exercise 2

In [204]:
lyrics = "Lean on me, when you're not strong\nAnd I'll be your friend\nI'll help you carry on"

In [205]:
print(lyrics)

Lean on me, when you're not strong
And I'll be your friend
I'll help you carry on


<br>Run the two lines of code above to see the string we'll be working with.

<br>Use string indexing to write code to print each line of the song on its own:

In [211]:
first_line = lyrics[:34]
print(first_line)

Lean on me, when you're not strong


In [214]:
second_line = lyrics[35:58]
print(second_line)

And I'll be your friend


In [215]:
third_line = lyrics[59:]
print(third_line)

I'll help you carry on


<br>Write code to replace the new line characters (`\n`) in `lyrics` with a period and a space ". ":

In [216]:
lyrics_on_one_line = lyrics.replace("\n", ". ")
print(lyrics_on_one_line)

Lean on me, when you're not strong. And I'll be your friend. I'll help you carry on


<br>Write code to capitalize the first letter of every word in `lyrics`:

In [217]:
lyrics.title()

"Lean On Me, When You'Re Not Strong\nAnd I'Ll Be Your Friend\nI'Ll Help You Carry On"

<br><br>*If this result looks strange to you, we will learn how to fix it tomorrow!*

### <br><br><span style="color:red">Strings: Secret Message Exercise

For this exercise, you are going to transform one string into another. First I will show you an example. Then you will write one of your own.

In [218]:
full_message = "I ate so many cookies now I feel sick."
print(full_message)

I ate so many cookies now I feel sick.


In [219]:
full_message[:-12]

'I ate so many cookies now '

In [220]:
full_message[:-12].replace("ate so", "want")

'I want many cookies now '

In [221]:
full_message[:-12].replace("ate so", "want").replace("any", "ore")

'I want more cookies now '

In [222]:
full_message[:-12].replace("ate so", "want").replace("any", "ore").title()

'I Want More Cookies Now '

In [223]:
full_message[:-12].replace("ate so", "want").replace("any", "ore").title().replace(" ", ". ")

'I. Want. More. Cookies. Now. '

In [224]:
full_message[:-12].replace("ate so", "want").replace("any", "ore").title().replace(" ", ". ") + "Please."

'I. Want. More. Cookies. Now. Please.'

<br><br>**<span style="color:crimson">LOGIC.** Now it is your turn. Think of a string that you want to turn into a different string. Use trial and error to build up a line of code that will transform your message.

<br>You can use indexing to take a substring, any of the string functions you've learned today (`rstrip()`, `lstrip()`, `replace()`, `upper()`, `lower()`, `capitalize()`, `title()`), as well as the addition operator `+` to add new characters at the end or beginning. Include at least 5 steps to get from your full message to your secret message.

In [ ]:
full_message = ""

In [225]:
full_message = "chicago is a city in illinois, usa"

In [226]:
(full_message.replace("chicago", "edinburgh").replace("illinois", "scotland").replace("usa", "uk").title().replace("A ", "a great ")[:-2]+"UK").lstrip("Edinburgh Is ").capitalize()

'A great city in scotland, uk'

In [227]:
full_message = "Summer is the best season in Chicago"

In [228]:
("What ".upper() + full_message[full_message.find("is"):].replace("Chicago", "Los Angeles").upper() + "?").replace(" ", "-")

'WHAT-IS-THE-BEST-SEASON-IN-LOS-ANGELES?'

In [229]:
full_message = "(She Believes)"

In [231]:
full_message.replace("es", "ed").rstrip(")").lstrip("(").lower()[1:].replace("be", "")[:-3] + "d"





'he lied'

In [233]:
full_message = "I don't feel so good today."

In [234]:
full_message[:-7].replace("I don't", "I'm").replace("feel", "feeling").replace("so","pretty, "*3).replace(" good","good!")

"I'm feeling pretty, pretty, pretty, good!"

## <br><br><br>Day 1 Quiz

On your own, before we meet again for tomorrow's morning lecture, complete the Day 1 Quiz. The Jupyter Notebook file is called "day1Quiz.ipynb" and is in the same folder as this notebook. If you are using Colab, go to <https://colab.research.google.com/github/aGitHasNoName/pythonBootcamp_3Day/blob/main/day1Quiz.ipynb>.

<br>If you have questions about the quiz material, send me an email and I will try my best to reply to all emails before tomorrow's lecture (which starts at 10 am Central on the same Zoom channel as today).
<br><br>As a reminder, the quiz is self-graded - you do not need to turn anything in! The answer key is called "day1Quiz-answers.ipynb". For Colab users, the answer key is at <https://colab.research.google.com/github/aGitHasNoName/pythonBootcamp_3Day/blob/main/day1Quiz-answers.ipynb>.